common/djangoapps/config_models/models.py

```diff
@@ -22,46 +23,105 @@ class ConfigurationModel(models.Model):
-        return 'configuration/{}/current'.format(cls.__name__)
+        if cls.KEY_FIELDS != ():
+            if len(args) != len(cls.KEY_FIELDS):
+                raise TypeError(
+                    "cache_key_name() takes exactly {} arguments ({} given)".format(len(cls.KEY_FIELDS), len(args))
+                )
+            return u'configuration/{}/current/{}'.format(cls.__name__, ','.join(unicode(arg) for arg in args))
+        else:
+            return 'configuration/{}/current'.format(cls.__name__)
 
     @classmethod
-    def current(cls):
+    def current(cls, *args):
         """
         Return the active configuration entry, either from cache,
         from the database, or by creating a new empty entry (which is not
         persisted).
         """
-        cached = cache.get(cls.cache_key_name())
+        cached = cache.get(cls.cache_key_name(*args))
         if cached is not None:
             return cached
 
+        key_dict = dict(zip(cls.KEY_FIELDS, args))
         try:
-            current = cls.objects.order_by('-change_date')[0]
+            current = cls.objects.filter(**key_dict).order_by('-change_date')[0]
         except IndexError:
-            current = cls()
+            current = cls(**key_dict)
 
-        cache.set(cls.cache_key_name(), current, cls.cache_timeout)
+        cache.set(cls.cache_key_name(*args), current, cls.cache_timeout)
         return current
 
     @classmethod
+    def current_set(cls):
+        """
+        A queryset for the active configuration entries only. Only useful if KEY_FIELDS is set.
+        """
+        assert cls.KEY_FIELDS != (), "Just use .current() if there are no KEY_FIELDS"
+        return cls.objects.extra(where=["id IN ({subquery})".format(subquery=cls._current_ids_subquery())])
+
+    @classmethod
+    def _current_ids_subquery(cls):
+        """
+        Internal helper method to return an SQL string that will get the IDs of
+        all the current entries (i.e. the most recent entry for each unique set
+        of key values). Only useful if KEY_FIELDS is set.
+        """
+        key_fields_escaped = [connection.ops.quote_name(name) for name in cls.KEY_FIELDS]
+        # The following assumes that the rows with the most recent date also have the highest IDs
+        return "SELECT MAX(id) FROM {table_name} GROUP BY {key_fields}".format(
+            key_fields=', '.join(key_fields_escaped), table_name=cls._meta.db_table
+        )
+
+    @classmethod
     def is_enabled(cls):
         """Returns True if this feature is configured as enabled, else False."""
         return cls.current().enabled
+
+    @classmethod
+    def key_values_cache_key_name(cls, *key_fields):
+        """ Key for fetching unique key values from the cache """
+        key_fields = key_fields or cls.KEY_FIELDS
+        return 'configuration/{}/key_values/{}'.format(cls.__name__, ','.join(key_fields))
+
+    @classmethod
+    def key_values(cls, *key_fields, **kwargs):
```

Comment:
Why does this need to take a list of `key_fields`? Should that list ever not be `cls.KEY_FIELDS`?